### PR TITLE
Do not try to take QEMU snapshots with virtio-gpu

### PR DESCRIPTION
**Pull request by Vogtinator**
virtio-gpu does not appear to support serialization yet.

### CODE REVIEW

**Comment by foursixnine** [comment]
Only the aarch64 workers will be hit by this, but on the other hand it would render test developers unable to use test snapshots to speedup development of test which is a trade off

**Comment by Vogtinator** [comment]
> Only the aarch64 workers will be hit by this, but on the other hand it would render test developers unable to use test snapshots to speedup development of test which is a trade off

This should not have any negative impact at all, as currently using QEMUVGA=virtio just causes every test to fail due to savevm aborting:

```
State blocked by non-migratable device '0000:00:02.0/virtio-gpu'
DIE Could not save snapshot 'lastgood': savevm lastgood
State blocked by non-migratable device '0000:00:02.0/virtio-gpu' at /usr/lib/os-autoinst/backend/qemu.pm line 241.
```

Or this is not the case on aarch64?

**Comment by foursixnine** [comment]
That sounds like a kind of old version of qemu to me... Looks like you're being hit by bsc#1025025

**Comment by Vogtinator** [comment]
> Looks like you're being hit by bsc#1025025

Indeed, that's probably it. I assume that's available on o3 as well?

**Comment by foursixnine** [comment]
Yes, all of the aarch64 workers are using Qemu 2.9.x

**Comment by Vogtinator** [comment]
This is mainly about x86_64, I guess that's the case for those workers as well?

**Comment by foursixnine** [comment]
Indeed, they are running 2.6.2 on x86_64

**Comment by okurz** [comment]
LGTM anyway

File: backend/qemu.pm [hunk 1]
```diff
@@ -175,9 +175,17 @@ sub can_handle {
     my ($self, $args) = @_;
     my $vars = \%bmwqemu::vars;
 
-    # XXX: Temporary (hopefully) workaround for nvme, since snapshots fails.
-    # See also https://github.com/os-autoinst/os-autoinst/pull/781
-    if ($args->{function} eq 'snapshots' && $vars->{HDDFORMAT} ne 'raw' && $vars->{HDDMODEL} ne 'nvme') {
+    if ($args->{function} eq 'snapshots') {
+        # raw HDD format does not support snapshots
+        return unless $vars->{HDDFORMAT} ne 'raw';
```
Comment (coolo) [inline backend/qemu.pm]: Avoid double negation

Comment (Vogtinator) [inline backend/qemu.pm]: Done.

File: backend/qemu.pm [hunk 2]
```diff
@@ -175,9 +175,17 @@ sub can_handle {
...
 
-    # XXX: Temporary (hopefully) workaround for nvme, since snapshots fails.
-    # See also https://github.com/os-autoinst/os-autoinst/pull/781
-    if ($args->{function} eq 'snapshots' && $vars->{HDDFORMAT} ne 'raw' && $vars->{HDDMODEL} ne 'nvme') {
+    if ($args->{function} eq 'snapshots') {
+        # raw HDD format does not support snapshots
+        return if $vars->{HDDFORMAT} eq 'raw';
+
+        # virtio-gpu does not support saving yet
+        return if $vars->{QEMUVGA} eq 'virtio';
```
Comment (coolo) [inline backend/qemu.pm]: if I read the code right QEMUVGA might be undef - this will give you a warning then.

Comment (Vogtinator) [inline backend/qemu.pm]: Fixed. I did not test my fix though.

File: backend/qemu.pm [hunk 3]
```diff
@@ -180,7 +180,7 @@ sub can_handle {
         return if $vars->{HDDFORMAT} eq 'raw';
 
         # virtio-gpu does not support saving yet
-        return if $vars->{QEMUVGA} eq 'virtio';
+        return if defined?($vars->{QEMUVGA}) && $vars->{QEMUVGA} eq 'virtio';
```
Comment (okurz) [inline backend/qemu.pm]: that looks weird. It's intended to look like `defined?(…)`?

Comment (Vogtinator) [inline backend/qemu.pm]: Whoops, I totally mixed up ruby and perl there.

Comment (coolo) [inline backend/qemu.pm]: I prefer 
```
return if ($vars->{QEMUVGA} || '') eq 'virtio';
```

**Comment by coolo** [comment]
So now please squash.

**Comment by Vogtinator** [comment]
Can't you do that with the GitHub PR merge button directly?

**Comment by coolo** [comment]
never tried - I only squash if there is just one commit. But I'll try :)

**Pull request merged**